actionpack/lib/action_dispatch/journey/matched_route.rb

```diff
@@ -0,0 +1,107 @@
+# frozen_string_literal: true
+
+module ActionDispatch
+  module Journey # :nodoc:
+    class MatchedRoute # :nodoc:
+      delegate :defaults, :required_parts, :grouped_optional_parts, to: :@route
+
+      def initialize(route, constraints, options, path_parameters, parameterize)
+        @route = route
+        @constraints = constraints
+        @options = options
+        @path_parameters = path_parameters
+        @parameterize = parameterize
+      end
+
+      def parameterized_parts
+        @parameterized_parts ||= @constraints.dup.tap do |parameterized_parts|
+          keys_to_keep = @route.parts.reverse_each.drop_while { |part|
+            !@options.key?(part) || (@options[part] || @path_parameters[part]).nil?
+          } | required_parts
+
+          parameterized_parts.delete_if do |bad_key, _|
+            !keys_to_keep.include?(bad_key)
+          end
+
+          if @parameterize
+            parameterized_parts.each do |k, v|
+              parameterized_parts[k] = @parameterize.call(k, v)
+            end
+          end
+
+          parameterized_parts.keep_if { |_, v| v }
+        end
+      end
+
+      def tests
+        @tests ||= @route.path.requirements
+      end
+
+      module RegexCaseComparator
+        DEFAULT_INPUT = /[-_.a-zA-Z0-9]+\/[-_.a-zA-Z0-9]+/
+        DEFAULT_REGEX = /\A#{DEFAULT_INPUT}\Z/
+
+        def self.===(regex)
+          DEFAULT_INPUT == regex
+        end
+      end
+
+      def missing_keys
+        @missing_keys ||= required_parts.each_with_object([]) do |key, memo|
+          case tests[key]
+          when nil
+            unless parameterized_parts[key]
+              memo << key
+            end
+          when RegexCaseComparator
+            unless RegexCaseComparator::DEFAULT_REGEX === parameterized_parts[key]
+              memo << key
+            end
+          else
+            unless /\A#{tests[key]}\Z/ === parameterized_parts[key]
+              memo << key
+            end
+          end
+        end
+      end
+
+      def invalid?
+        !missing_keys.empty?
+      end
+
+      def params
+        @options.reject do |key, _|
+          parameterized_parts.key?(key) || defaults.key?(key)
+        end
+      end
+
+      def formatted_route
+        parts = parameterized_parts.dup
```

Comment:
I doubt it amounts to much, but parameterized parts is allocated twice due to this line. I think the code can probably be reordered to only allocate it once. But I don't know if the result would be measurably faster.

One thing that could help with reasoning about this is to mark public and private methods, to make it clearer what is and what isn't called by the use place of the class.

Reply:
Thanks @yxhuvud, I've tried that but results were pretty much the same - `1.17`x slower.
Plus, one test was red.